File: AndroidBasics/calculator/src/main/java/pan/alexander/calculator/di/RoomModule.java
package pan.alexander.calculator.di;

import android.app.Application;

import androidx.annotation.NonNull;
import androidx.room.Room;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import pan.alexander.calculator.data.database.AppDatabase;
import pan.alexander.calculator.data.database.HistoryDao;

@Module
public class RoomModule {
    private final AppDatabase db;

    public RoomModule(@NonNull Application appContext) {
        db = Room
                .databaseBuilder(appContext, AppDatabase.class, "app-database")
                .fallbackToDestructiveMigration()
                .build();
    }

    @Provides
    @Singleton
    AppDatabase providesRoomDatabase() {
        return db;
    }

    @Provides
    @Singleton
    HistoryDao providesHistoryDao() {
        return db.historyDao();
    }
}

File: simlpe-weather/app/src/main/java/pan/alexander/simpleweather/di/RoomModule.kt
package pan.alexander.simpleweather.di

import android.app.Application
import androidx.room.Room
import androidx.room.migration.Migration
import androidx.sqlite.db.SupportSQLiteDatabase
import dagger.Module
import dagger.Provides
import pan.alexander.simpleweather.data.database.AppDatabase
import pan.alexander.simpleweather.data.database.CurrentWeatherDao
import javax.inject.Singleton

@Module
class RoomModule(appContext: Application) {

    val MIGRATION_1_2 = object : Migration(1, 2) {
        override fun migrate(database: SupportSQLiteDatabase) {
            database.execSQL("ALTER TABLE CurrentWeather ADD COLUMN rain INTEGER DEFAULT -1 NOT NULL")
            database.execSQL("ALTER TABLE CurrentWeather ADD COLUMN snow INTEGER DEFAULT -1 NOT NULL")
        }
    }

    private val appDatabase = Room
        .databaseBuilder(appContext, AppDatabase::class.java, "main_database")
        .addMigrations(MIGRATION_1_2)
        .build()

    @Provides
    @Singleton
    fun providesRoomDatabase(): AppDatabase = appDatabase

    @Provides
    @Singleton
    fun providesCurrentWeatherDao(): CurrentWeatherDao = appDatabase.currentWeatherDao()
}
